cmd/gen-go-sample/inittree.go

```diff
@@ -114,19 +114,23 @@ func (t *initTree) get(k string, info pbinfo.Info) (*initTree, error) {
 				if d.GetOptions().GetMapEntry() == true {
 					// type is a map entry
 					for _, f2 := range d.Field {
-						if f2.GetName() == "key" {
+						switch f2.GetName() {
+						case "key":
 							// key is always a primitive type
 							v.typ.keyType = &initType{prim: f2.GetType()}
-						}
-						if f2.GetName() == "value" {
-							if tn2 := f.GetTypeName(); tn2 == "" {
-								// value is a primitive type
-								v.typ.valueType = &initType{prim: f2.GetType()}
-							} else if typ := info.Type[tn2]; typ == nil {
-								return nil, errors.E(nil, "cannot find descriptor of %q", f.GetTypeName())
+						case "value":
+							if t2 := f2.GetType(); t2 == descriptor.FieldDescriptorProto_TYPE_MESSAGE || t2 == descriptor.FieldDescriptorProto_TYPE_ENUM {
+								// value is a protobuf message or enum
+								tn2 := f2.GetTypeName()
+								typ2 := info.Type[tn2]
+								if typ2 == nil {
+									return nil, errors.E(nil, "cannot find descriptor of %q", tn2)
+								}
+								
+								v.typ.valueType = &initType{desc: typ2}
 							} else {
```

Comment:
So WRT to using an `else` statement here, I tend to take the conventions described in [Effective Go](https://golang.org/doc/effective_go.html#if) beyond `error` cases. What I mean by that is that the **basic** flow of the code should *not* be in an `else` and special cases should be handled by `if`/`else if`. This is why in my suggested snippet for this section I had something like

```go
v.typ.valueType = &initType{prim: t2}

if t2 := f2.GetType(); t2 == Message ...
  ...
  v.typ.valueType = &initType{desc: typ2}
}
```

The primitive type case becomes the basic case (because there is no extra work necessary other than assignment) and the special case of a Message or Enum type becomes handled specifically (it has more involved logic). I think this is easier to read because I don't need to go back up to the original `if` to understand what was being evaluated to get me in the `else`, it reads straight down. Combined with the comment you added `// value is a primitive type` I think this is much easier to read.

Sorry for being so nit-picky and if you feel this is unnecessary feel free to say "thanks, but no thanks".

Reply:
Thanks and thanks :)

I didn't know this is the convention adopted by Go, and therefore I completely missed this difference in your code snippet.

Updated :)